File: adapters/node/router.ts
import fs from "node:fs"; import yaml from "js-yaml";
type Inp = any; type Tool = {name:string;module:string;output:string;detect:any};
const reg = yaml.load(fs.readFileSync("tool_definitions/q4.tools.yaml","utf8")) as any;
const routing = yaml.load(fs.readFileSync("core/q4.router.yaml","utf8")) as any;
const tools: Tool[] = reg.tools;

const hasBuckets = (inp:Inp)=> !!inp.buckets && ["S","W","O","T"].every(k=>Array.isArray(inp.buckets[k])&&inp.buckets[k].length>0);
const coverage = (items:any[], req:string[]) => items.length===0?0:items.filter((it)=>req.every(f=>f in it)).length/items.length;

export function routeTool(inp:Inp): Tool | null {
  if (inp.mode==="force" && inp.framework_hint) { return tools.find(t=>t.name===String(inp.framework_hint).toLowerCase())||null; }
  if (hasBuckets(inp)) { return tools.find(t=>t.name==="swot")||null; }
  const items = inp.items||[];
  let best:Tool|null=null; let bestScore=-1;
  for (const t of tools) {
    const req=t.detect?.required_fields;
    if (!req) continue;
    const sc=coverage(items, req);
    if (sc>bestScore) {best=t; bestScore=sc;}
  }
  if (best && bestScore>=routing.routing.thresholds.coverage) return best;
  const hint = new Set((inp.axes_hint||[]).map((s:string)=>s.toLowerCase()));
  if (hint.size) return tools.find(t=>hint.has(String(t.detect?.axes?.x))||hint.has(String(t.detect?.axes?.y)))||null;
  return null;
}

File: adapters/python/router.py
import yaml, json
from typing import Any, Dict, List

def load_yaml(p): return yaml.safe_load(open(p,"r",encoding="utf-8"))

REG = load_yaml("tool_definitions/q4.tools.yaml")
ROUT = load_yaml("core/q4.router.yaml")

def coverage(items: List[Dict], req: List[str])->float:
    if not items: return 0.0
    return sum(1 for it in items if all(f in it for f in req))/len(items)

def route_tool(inp: Dict[str,Any]) -> Dict[str,Any] | None:
    tools: List[Dict] = REG["tools"]
    if inp.get("mode")=="force" and inp.get("framework_hint"):
        for t in tools:
            if t["name"] == str(inp["framework_hint"]).lower(): return t
    b = inp.get("buckets") or {}
    if all(k in b and isinstance(b[k], list) and len(b[k])>0 for k in ["S","W","O","T"]):
        return next(t for t in tools if t["name"]=="swot")
    items = inp.get("items") or []
    best=None; best_score=-1.0
    for t in tools:
        req = t.get("detect",{}).get("required_fields")
        if not req: continue
        sc = coverage(items, req)
        if sc>best_score: best=t; best_score=sc
    if best and best_score >= ROUT["routing"]["thresholds"]["coverage"]: return best
    hint = {s.lower() for s in (inp.get("axes_hint") or [])}
    if hint:
        for t in tools:
            ax = (t.get("detect",{}).get("axes") or {})
            if {str(ax.get("x","")).lower(), str(ax.get("y","")).lower()} & hint: return t
    return None